### ISSUE TITLE
Input to spanning_tree

### ISSUE THREAD

**Issue by chijames** [choidami/sst#3]
Hi,

Can you explain a little bit about why https://github.com/choidami/sst/blob/main/core/spanning_tree.py#L43 takes logits of shape n x (n-1) ? As far as I understand, section 3 of [koo et al 07] needs to calculate the root embedding for both single and multi root cases. In other words, why isn't the shape being n x n? (where the diagonal entries account for root scores)

Thanks a lot!

**Comment by chijames** [comment]
Ok, I think I know the reason.

It's because we don't actually care about roots(heads) of the NRI task. But for the Listops task, we do need to consider root scores. That is also why the Torch-Struct library is needed, isn't it? Also, are you using DependencyCRF or NonProjectiveDependencyCRF?

Thanks.

**Comment by choidami** [comment]
Hello,

You're exactly right- for the NRI task, since we're only interested in undirected spanning trees, there is no concept of a root. Though when we defined the weights matrix W (https://github.com/choidami/sst/blob/main/core/spanning_tree.py#L52) we actually meant for the diagonals to mean self loops, which spanning trees never have, which is why logits is n x (n - 1).

For the ListOps task, we assume the 0th token is the root node, so we construct a (n-1) x (n-1) matrix where the diagonal corresponds to the logits from 0 to the other nodes. We used non-projective.

Dami
